### ISSUE TITLE
Define default logical optimizer

### ISSUE THREAD

**Issue by hartig** [LiUSemWeb/HeFQUIN#347]
We have a heuristics-based logical optimizer with a number of heuristics, where most of them are currently commented out. Decide on a selection of heuristics that works okay for several example queries even without any physical optimizer and define this selection as the logical optimizer that HeFQUIN uses by default.

**Comment by keski** [comment]
Using the Fedshop queries as a reference, the following optimizations are useful (in principle) for several the FedShop queries:
- `MergeRequests`
- `FilterPushDown`
- `GreedyBasedReordering`
- `PushJoinUnderUnionWithRequests`

These are low cost optimizations that should probably be enabled by default.

`CardinalityBasedJoinOrderingWithRequests` is less straightforward. Many of the FedShop queries only use unions (no joins between requests), and the cost associated with getting the cardinalities is pretty high. I think this should be tested experimentally we add it to the default optimizer.

The following are not required at all for the FedShop queries:
- `ApplyVocabularyMappings`
- `RemoveUnnecessaryL2gAndG2l`
- `PullUpLtgOverUnion`
- `PullUpLtgOverJoin`
- `RemovePairsOfG2lAndL2g`

**Comment by hartig** [comment]
> Using the Fedshop queries as a reference, the following optimizations are useful (in principle) for several the FedShop queries:
>     * `MergeRequests`
>     * `FilterPushDown`
>     * `GreedyBasedReordering`
>     * `PushJoinUnderUnionWithRequests`
> These are low cost optimizations that should probably be enabled by default.

Yes. But notice that `GreedyBasedReordering` is actually unnecessary if we run `CardinalityBasedJoinOrderingWithRequests` afterwards (and vice versa).

> `CardinalityBasedJoinOrderingWithRequests` is less straightforward. Many of the FedShop queries only use unions (no joins between requests), and the cost associated with getting the cardinalities is pretty high. I think this should be tested experimentally we add it to the default optimizer.

Yes. `CardinalityBasedJoinOrderingWithRequests` is useless for queries that do not contain any joins (except for joins within requests). However, an easy way to avoid the issue is to extend the implementation of `CardinalityBasedJoinOrderingWithRequests` (or, its superclass, to be more precise) such that it first checks whether the given logical plan contains join operators (`LogicalOpJoin` or `LogicalOpMultiwayJoin`) and, if that is not the case, then the given logical plan is simply returned without attempting to reorder anything. This check would have to be added before line 73 in `CardinalityBasedJoinOrderingBase`.

> The following are not required at all for the FedShop queries:
>     * `ApplyVocabularyMappings`
>     * `RemoveUnnecessaryL2gAndG2l`
>     * `PullUpLtgOverUnion`
>     * `PullUpLtgOverJoin`
>     * `RemovePairsOfG2lAndL2g`

Obviously. These are all related to applying vocabulary mappings (l2g and g2l are the logical operators that apply a vocabulary mapping to the solution mappings in intermediate results). In principle, they can be disabled if none of the federation members is associated with a vocabulary mapping. On the other hand, they should also be low cost, especially for plans that neither contain l2g and g2l nor contain requests to fed.members with a vocab.mapping.

**Comment by keski** [comment]
> Agreed, avoiding it would definitely be easy and something we should add. I'll create a PR for it and reference this issue.

I created a sub-issue (https://github.com/LiUSemWeb/HeFQUIN/issues/465) and a PR (https://github.com/LiUSemWeb/HeFQUIN/pull/466).
